src/Framework/ExtendedBuildWarningEventArgs.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Framework;
+
+/// <summary>
+/// Generic custom warning events including extended data for event enriching.
+/// Extended data are implemented by <see cref="IExtendedBuildEventArgs"/>
+/// </summary>
+public sealed class ExtendedBuildWarningEventArgs : BuildWarningEventArgs, IExtendedBuildEventArgs
```

Comment:
This can probably get converted (and so custom data thrown away) by our code in some places - e.g. https://github.com/dotnet/msbuild/blob/main/src/Build/BackEnd/Components/Logging/LoggingService.cs#L1437-L1478
It is not a new situation - same would happen even before, but as we are now explicitly guiding to use those new events - we should probably at least document that in xml comments

Reply:
Yes, good catch. I am aware of it, this is however same for today customer errors and messages derived from existing EventArgs.
However, this is probably still bug and now, we actually can fix it by reusing extended data. 
Will try...